implementations/micrometer-registry-timestream2/src/main/java/io/micrometer/timestream2/TimestreamMeterRegistry.java

```diff
@@ -0,0 +1,333 @@
+ * @since 1.6.0
+ */
+public class TimestreamMeterRegistry extends StepMeterRegistry {
+
+    static final ThreadFactory DEFAULT_THREAD_FACTORY = new NamedThreadFactory("timestream-metrics-publisher");
+
+    private static final WarnThenDebugLogger warnThenDebugLogger = new WarnThenDebugLogger(TimestreamMeterRegistry.class);
+
+    private final Logger logger = LoggerFactory.getLogger(TimestreamMeterRegistry.class);
+
+    private final TimestreamConfig config;
+
+    private final TimestreamWriteAsyncClient timestreamWriteAsyncClient;
+
+    /**
+     * Create a new instance with given parameters.
+     *
+     * @param config        configuration to use
+     * @param clock         clock to use
+     * @param threadFactory thread factory to use
+     * @param client        TimeStream client to use
+     */
+    //VisibleForTesting
+    TimestreamMeterRegistry(TimestreamConfig config, Clock clock, ThreadFactory threadFactory, TimestreamWriteAsyncClient client) {
+        super(config, clock);
+        config().namingConvention(new TimestreamNamingConvention());
+        this.config = config;
+        checkClient(client);
+        this.timestreamWriteAsyncClient = client;
+        start(threadFactory);
+    }
+
+    private void checkClient(TimestreamWriteAsyncClient client) {
+        Objects.requireNonNull(client);
+    }
+
+    @Override
+    protected void publish() {
+        MeterPartition.partition(this, config.batchSize())
+                .stream()
+                .map(this::toRecords)
+                .forEach(this::sendRecords);
+    }
+
+    //VisibleForTesting
+    List<Record> toRecords(List<Meter> partition) {
+        Batch batch = new Batch();
+        return partition.stream().flatMap(m -> m.match(
+                batch::gaugeData,
+                batch::counterData,
+                batch::timerData,
+                batch::summaryData,
+                batch::longTaskTimerData,
+                batch::timeGaugeData,
+                batch::functionCounterData,
+                batch::functionTimerData,
+                batch::metricData)
+        ).collect(toList());
+    }
+
+    //VisibleForTesting
+    void sendRecords(List<Record> records) {
+        WriteRecordsRequest writeRecordsRequest = WriteRecordsRequest.builder()
+                .databaseName(config.databaseName())
+                .tableName(config.tableName())
+                .records(records)
+                .build();
+
+        timestreamWriteAsyncClient
+                .writeRecords(writeRecordsRequest)
+                .thenRun(() ->
+                        logger.debug("published records to database: {}, table: {} ", writeRecordsRequest.databaseName(), writeRecordsRequest.tableName())
+                )
+                .exceptionally(throwable -> {
+                    if (throwable instanceof AbortedException) {
+                        logger.warn("sending records data was aborted: " + throwable.getMessage(), throwable);
```

Comment:
This way, the message will be duplicated, e.g. if you do something like this:

```
Exception exception = new IllegalArgumentException("This is the message of a simulated exception.");
logger.info("Something bad happened: " + exception.getMessage(), exception);
```

The output will look like this

```
18:32:37.536 [main] INFO io.micrometer.core.Main - Something bad happened: This is the message of a simulated exception.
java.lang.IllegalArgumentException: This is the message of a simulated exception.
	at io.micrometer.core.Main.main(Main.java:10)
```

Since the message of the `Throwable` will be logged out too, I don't think adding the message is necessary, I would just do : `logger.warn("sending records data was aborted", throwable);` like you did in the error case.

Reply:
you're right